service/src/LevelZero.cpp

```diff
@@ -523,6 +526,366 @@ namespace geopm
+//                            ": LevelZero Metric Context Deactivation failed",
+//                            __LINE__);
+//        }
+    }
+
+
+    void LevelZeroImp::metric_init(unsigned int l0_device_idx, unsigned int l0_domain_idx)
+    {
+        ze_result_t ze_result;
+        ze_context_handle_t context = m_devices.at(l0_device_idx).subdevice.context.at(l0_domain_idx);
+        ze_result = zetContextActivateMetricGroups(context, m_devices.at(l0_device_idx).subdevice_handle.at(l0_domain_idx),
+                                                   1, &m_devices.at(l0_device_idx).subdevice.metric_group_handle.at(l0_domain_idx));
+
+        ze_event_pool_handle_t event_pool_handle = nullptr;
+        ze_event_pool_desc_t event_pool_desc = {ZE_STRUCTURE_TYPE_EVENT_POOL_DESC, nullptr, 0, 1};
+
+        ze_result = zeEventPoolCreate(context, &event_pool_desc, 1,
+                                      &m_devices.at(l0_device_idx).subdevice_handle.at(l0_domain_idx),
+                                      &event_pool_handle);
+
+        check_ze_result(ze_result, GEOPM_ERROR_RUNTIME,
+                        "LevelZero::" + std::string(__func__) +
+                        ": LevelZero Event Pool Create failed",
+                        __LINE__);
+        m_devices.at(l0_device_idx).subdevice.event_pool.push_back(event_pool_handle);
+
+        ze_event_desc_t event_desc = {ZE_STRUCTURE_TYPE_EVENT_DESC, nullptr, 0,
+                                      ZE_EVENT_SCOPE_FLAG_HOST, ZE_EVENT_SCOPE_FLAG_HOST};
+
+        ze_event_handle_t event = nullptr;
+        ze_result = zeEventCreate(event_pool_handle, &event_desc, &event);
+        check_ze_result(ze_result, GEOPM_ERROR_RUNTIME,
+                        "LevelZero::" + std::string(__func__) +
+                        ": LevelZero Event Create failed",
+                        __LINE__);
+
+        m_devices.at(l0_device_idx).subdevice.event.push_back(event);
+
+        zet_metric_streamer_desc_t metric_streamer_desc = {
+            ZET_STRUCTURE_TYPE_METRIC_STREAMER_DESC,
+            nullptr,
+            32768,
+            m_devices.at(l0_device_idx).metric_sampling_period};
+        zet_metric_streamer_handle_t metric_streamer = nullptr;
+
+        ze_result = zetMetricStreamerOpen(context, m_devices.at(l0_device_idx).subdevice_handle.at(l0_domain_idx), m_devices.at(l0_device_idx).subdevice.metric_group_handle.at(l0_domain_idx), &metric_streamer_desc, event, &metric_streamer);
+
+        check_ze_result(ze_result, GEOPM_ERROR_RUNTIME,
+                        "LevelZero::" + std::string(__func__) +
+                        ": LevelZero Metric Streamer Open failed",
+                        __LINE__);
+
+        m_devices.at(l0_device_idx).subdevice.metric_streamer.push_back(metric_streamer);
+    }
+
+    void LevelZeroImp::metric_calc(unsigned int l0_device_idx, unsigned int l0_domain_idx,
+                                   zet_metric_streamer_handle_t metric_streamer) const
+    {
```

Comment:
Does this method require `metric_init()` and `metric_group_cache()` to already be called? If so, I would recommend adding an assertion (`GEOPM_DEBUG_ASSERT`) to ensure this.